BE-011 Hanna Kalashnik User coupon favorite+sorting

Hanna-Kalashnyk opened the pull request:
BE-011 Hanna Kalashnik User-Favorite-Coupon + sorting

Hanna-Kalashnyk committed:
User, favorite, coupon

Hanna-Kalashnyk committed:
+ findByEmail endpoint

Hanna-Kalashnyk committed:
+new PR

Hanna-Kalashnyk committed:
+ path import com.exadel.discount.model.security.Role;

Hanna-Kalashnyk committed:
+ userRepository.findByEmail

Hanna-Kalashnyk committed:
+ swagger description in controllers
+ slf4j logs in service layer
+ NotFoundException in service layer

Hanna-Kalashnyk committed:
- duplicate of findAllFavorites-method is deleted

Hanna-Kalashnyk committed:
Merge branch 'dev' into Coupon_Favourite_User

Hanna-Kalashnyk committed:
new ECF branch

el-dub commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+    @ApiOperation("Get list of all favorites with sorting")
+    /**  Get list of all users with sorting by params
+     sortDirection - ASC or DSC (unsorted - by default) ;
+     sortField - name of sorted field by (id - by default) **/
+    public List<FavoriteDto> getAllCoupons(@RequestParam(value = "sortDirection", defaultValue = "") String sortDirection,
+                                         @RequestParam(value = "sortField", defaultValue = "id") String sortField){
```
Mistake in naming

Hanna-Kalashnyk replied:
Done. Thank U!

el-dub commented on src/main/java/com/exadel/discount/controller/CouponController.java:
```diff
@@ -0,0 +1,68 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Sort;
+import org.springframework.format.annotation.DateTimeFormat;
+import org.springframework.web.bind.annotation.*;
```
Remove wildcards, please

Hanna-Kalashnyk replied:
Done

el-dub commented on src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+import com.exadel.discount.dto.favorite.CreateFavoriteDto;
+import com.exadel.discount.dto.favorite.FavoriteDto;
+import com.exadel.discount.entity.Discount;
```
You don't have this

el-dub replied:
Just comment all parts with missing classes temporarily

Hanna-Kalashnyk replied:
But I will soon. Otherwise it will not work anyway

el-dub commented on src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+import com.exadel.discount.entity.User;
+import com.exadel.discount.exception.NotFoundException;
+import com.exadel.discount.mapper.FavoriteMapper;
+import com.exadel.discount.repository.DiscountRepository;
```
And this

Hanna-Kalashnyk replied:
Have locally and under PR

el-dub commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @GetMapping("/ofuser")
+    @ApiOperation("Get favorites of certain user")
+    public List<FavoriteDto> getFavoritesOfUser(@RequestParam("userId") final UUID userId) {
+        return favoriteService.getFavoritesOfUser(userId);
+    }
```
I think you should add sorting here as well

Hanna-Kalashnyk replied:
Will. A bit later

el-dub commented on src/main/java/com/exadel/discount/controller/CouponController.java:
```diff
@@ -0,0 +1,68 @@
+     sortField - name of sorted field by (id - by default)- date **/
+    public List<CouponDto> getAllCoupons(@RequestParam(value = "sortDirection", defaultValue = "") String sortDirection,
+                                         @RequestParam(value = "sortField", defaultValue = "id") String sortField){
+        Sort sort = Sort.unsorted();
+        if(!sortDirection.equals("")) {
+            sort = Sort.by(Sort.Direction.valueOf(sortDirection.toUpperCase()), sortField);
+        }
```
Maybe this part should be in service layer, not in controller

Hanna-Kalashnyk replied:
Done

el-dub commented on src/main/java/com/exadel/discount/repository/CouponRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+@Repository
+public interface CouponRepository extends JpaRepository<Coupon, UUID> {
+    <Optional> Coupon findCouponByDate(LocalDateTime date);
```
I still don't think that it's ritght syntax

Hanna-Kalashnyk replied:
It was mentor's correction about <Optional> - caused the exception ability

el-dub replied:
Optional "<"Coupon">" maybe?

Hanna-Kalashnyk replied:
Thank U!!! Sharp eye)

Hanna-Kalashnyk replied:
Done

Hanna-Kalashnyk committed:
+filtering to Users

Hanna-Kalashnyk committed:
- @Transactional from service layer

Hanna-Kalashnyk committed:
+Sorting removeng to Service layer

el-dub commented on src/main/java/com/exadel/discount/dto/user/UserCityDto.java:
```diff
@@ -0,0 +1,42 @@
+package com.exadel.discount.dto.user;
+
+import com.exadel.discount.entity.Country;
```
Here is no need in this import

Hanna-Kalashnyk replied:
Done

el-dub commented on src/main/java/com/exadel/discount/entity/User.java:
```diff
@@ -0,0 +1,75 @@
+    private List<Favorite> favorites = new ArrayList<>();
+
+    @ManyToOne(fetch = FetchType.EAGER)
+    @JoinColumn(name="city_id", nullable=false)
+    private City city;
```
App doesn't working because of the absence of some classes. It's very bad for dev, I think. Imagine your PR will be approved, somebody will have his own PR and will merge dev into his branch. He still hasn't theese classes and app doesn't work, he can't even test it

el-dub replied:
So it should be commented I think

Hanna-Kalashnyk replied:
My part does not work without it(

Hanna-Kalashnyk replied:
Done

el-dub commented on src/main/java/com/exadel/discount/mapper/UserCityMapper.java:
```diff
@@ -0,0 +1,24 @@
+    @Mapping(source = "user.phone", target = "phone")
+    @Mapping(source = "user.email", target = "email")
+    @Mapping(source = "city.name", target = "cityName")
+    @Mapping(source = "country.name", target = "country")
+    UserCityDto toUserCityDto(User user, City city, Country country);
+}
```
Doesn't work

el-dub replied:
Please, test just your branch with dev merged into it, it should work. Now there is a lot of compile errors

Hanna-Kalashnyk replied:
Merged. Works...

Hanna-Kalashnyk replied:
Done

el-dub commented on src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    public List<FavoriteDto> getFavoritesOfUser(UUID userId) {
+        log.debug("Finding Favorites of certain user");
+
+        List<FavoriteDto> FavoriteDtoList = favoriteMapper.toFavoriteDtoList(favoriteRepository.findAll()
+                .stream()
+                .filter(s -> s.getUser().getId().equals(userId))
+                .collect(Collectors.toList()));
```
You should do it in repository like Artem has said today. try to add method findByUser()

Hanna-Kalashnyk replied:
Done

Hanna-Kalashnyk committed:
_ Optional into Repository layer because we have exception handler
+ some typos

Hanna-Kalashnyk committed:
User sorting works well

Hanna-Kalashnyk committed:
cleaning

Hanna-Kalashnyk committed:
+ AllCoupon Sorting

Hanna-Kalashnyk committed:
Coupon filtering works

Hanna-Kalashnyk committed:
works

el-dub commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+     sortField - name of sorted field by (id - by default) **/
+    public List<FavoriteDto> getAllFavorites(@RequestParam(value = "sortDirection", defaultValue = "") String sortDirection,
+                                         @RequestParam(value = "sortField", defaultValue = "id") String sortField){
+        Sort sort = Sort.unsorted();
+        if(!sortDirection.equals("")) {
+            sort = Sort.by(Sort.Direction.valueOf(sortDirection.toUpperCase()), sortField);
+        }
```
Move this to service like in coupons, please

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @GetMapping("{id}")
+    @ApiOperation("Get favorite by ID")
+    public FavoriteDto getFavoriteById(@PathVariable final UUID id) {
```
add annotation @NotNull

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @GetMapping("/ofuser")
+    @ApiOperation("Get favorites of certain user")
+    public List<FavoriteDto> getFavoritesOfUser(@RequestParam("userId") final UUID userId) {
```
add annotation @NotNull

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/controller/UserController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @GetMapping("{id}")
+    @ApiOperation("Get user by ID")
+    public UserCityDto getUserById(@PathVariable final UUID id) {
```
@NotNull

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @PostMapping
+    @ApiOperation("Save new favorite")
+    public FavoriteDto addFavorite(@RequestBody final CreateFavoriteDto createFavoriteDto) {
```
may be (@RequestBody @Validated(Create.class)  final CreateFavoriteDto createFavoriteDto)

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/controller/UserController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @GetMapping("/name")
+    @ApiOperation("Get user by lastname and firstname")
+    public List<UserDto> getUsersByName(@RequestParam("lastname") String lastName,
+                                        @RequestParam("firstame") String firstName) {
```
Add @NotNull

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/dto/favorite/CreateFavoriteDto.java:
```diff
@@ -0,0 +1,18 @@
+package com.exadel.discount.dto.favorite;
+
+import lombok.*;
```
remove wildcards

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/controller/FavoriteController.java:
```diff
@@ -0,0 +1,59 @@
+import io.swagger.annotations.ApiOperation;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Sort;
+import org.springframework.web.bind.annotation.*;
```
remove wildcards

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/dto/favorite/FavoriteDto.java:
```diff
@@ -0,0 +1,17 @@
+package com.exadel.discount.dto.favorite;
+
+import com.sun.istack.NotNull;
+import lombok.*;
```
remove wildcards

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/dto/user/UserCityDto.java:
```diff
@@ -0,0 +1,42 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode
+@ToString
```
@Data includes Equals & toString

Hanna-Kalashnyk replied:
Done

mblack666 commented on src/main/java/com/exadel/discount/entity/User.java:
```diff
@@ -0,0 +1,75 @@
+    @Enumerated(EnumType.STRING)
+    @Column(columnDefinition = "role")
+    @Type(type = "user_role")
+    private Role role;
```
Role can be null?

Hanna-Kalashnyk replied:
Done

Hanna-Kalashnyk committed:
+ DiscontRepository class
+ add some notFound exceptions to impl classes at empty foundLists
+ Feedback handling

Hanna-Kalashnyk committed:
removed findCouponsBetweenDates

Hanna-Kalashnyk committed:
findAllFavorites -> Page

Artem-Kishenets reviewed (approved):
.findAll were deleted, so the Java Virtual Machine can sleep well

Hanna-Kalashnyk committed:
cleaning

Hanna-Kalashnyk committed:
unsuccessful attempt with Query(join fetch) and countQuery

Hanna-Kalashnyk committed:
SQLGrammarException about "fetch"

Hanna-Kalashnyk committed:
+ SortPageMaker class
+ all searches in repos
+ doc

Hanna-Kalashnyk committed:
+sorting works!
without soiv
role searching throws
    "message": "No User with role USER is found"

Hanna-Kalashnyk committed:
role works
